Default Message Encryptor Cipher to AES-256-GCM From AES-256-CBC

assain opened the pull request:
@kaspth 
Currently ActiveSupport::MessageEncryptor defaults to CBC encryption, but it could stand to benefit from defaulting to `aes-256-gcm` encryption for the same reasons as cookies.

kaspth commented on Gemfile.lock:
```diff
@@ -433,4 +433,4 @@ DEPENDENCIES
 
 BUNDLED WITH
-   1.14.6
+   1.15.0
```
Please remove these Gemfile.lock changes. See Eileen's contributing to Rails workshop around amending and rebasing for some help on how: https://youtu.be/7zoD6NZy6vY?t=1h25s

assain replied:
Thank you for the feedback @kaspth, I'll fix it and add another commit :+1:

kaspth replied:
Don't add another commit. Just rebase/amend this one :)

kaspth commented:
For some reason Travis didn't run your pull request. Try the rebase/amend then force push.

assain committed:
set message_encryptor default cipher to aes-256-gcm

- Introduce a method to select default cipher, and maintain backward compatibility

kaspth commented on activesupport/test/message_encryptor_test.rb:
```diff
@@ -119,7 +119,7 @@ def assert_not_decrypted(value)
 
     def assert_not_verified(value)
-      assert_raise(ActiveSupport::MessageVerifier::InvalidSignature) do
+      assert_raise(ActiveSupport::MessageEncryptor::InvalidMessage) do
```
Why is this change needed?

assain replied:
Previously the setup method used  **aes-256-cbc**  cipher while writing this test and had made use of Message Verifier. However, after playing around with the tests, I realized this test was failing because ActiveSupport::MessageVerifier::InvalidSignature was raised by message verifier, and the new default cipher - **aes-256-gcm** used NullVerifier.

kaspth commented:
Okay, now we'll need a similar setting to what the AEAD cookies has here: https://github.com/rails/rails/pull/28132/files#diff-7f4141d796a8aad409136b2b6dc774b0R15

E.g. that there's a new 5.2 config to default the encryptor to gcm:

```ruby
# Other configs in new_framework_defaults_5_2.rb…

# Default encrypted messages to use AES-256-GCM encryption.
# Rails.application.active_support.message_encryptor.use_authenticated_encryption = true
```

Then auto default that to true in `load_defaults` under `"5.2"`.

If we don't have that config, we break backwardscompatibility — that's what the test failures show.

kaspth commented on railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults_5_2.rb.tt:
```diff
@@ -13,3 +13,6 @@
 # Rails.application.config.action_dispatch.use_authenticated_cookie_encryption = true
+
+# Use AES 256 GCM authenticated encryption as default for encrypting messages.
+# Rails.application.config.active_support.message_encryptor.use_authenticated_cookie_encryption = true
```
Shouldn't say `use_authenticated_cookie_encryption` but just `use_authenticated_encryption` 😊

assain replied:
I had corrected that mistake, don't know how it crept it again. Apologies!

kaspth commented on railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults_5_2.rb.tt:
```diff
@@ -13,3 +13,6 @@
 # Use AES 256 GCM authenticated encryption for encrypted cookies.
 # Existing cookies will be converted on read then written with the new scheme.
 # Rails.application.config.action_dispatch.use_authenticated_cookie_encryption = true
```
Ideally, we should be able to remove this config and use the below one to infer it. But that's for another time.

cc @mikeycgto

kaspth commented on activesupport/lib/active_support/message_encryptor.rb:
```diff
@@ -45,10 +56,10 @@ class InvalidMessage < StandardError; end
     # bits. If you are using a user-entered secret, you can generate a suitable
     # key by using <tt>ActiveSupport::KeyGenerator</tt> or a similar key
-    # derivation function.
+    # derivation function
```
Keep the period.

kaspth commented on activesupport/lib/active_support/message_encryptor.rb:
```diff
@@ -66,7 +77,7 @@ def initialize(secret, *signature_key_or_options)
       @secret = secret
       @sign_secret = sign_secret
-      @cipher = options[:cipher] || DEFAULT_CIPHER
+      @cipher = options[:cipher] || MessageEncryptor.default_cipher
```
`self.class.default_cipher`

kaspth commented on activesupport/lib/active_support/message_encryptor.rb:
```diff
@@ -19,7 +19,18 @@ module ActiveSupport
   class MessageEncryptor
-    DEFAULT_CIPHER = "aes-256-cbc"
+    class << self
+      #use_authenticated_message_encryption is used to activate 'aes-256-gcm' as default cipher
```
This comment doesn't really add anything to me. Let's just document it elsewhere.

kaspth commented on railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults_5_2.rb.tt:
```diff
@@ -13,3 +13,6 @@
 # Existing cookies will be converted on read then written with the new scheme.
 # Rails.application.config.action_dispatch.use_authenticated_cookie_encryption = true
+
+# Use AES 256 GCM authenticated encryption as default for encrypting messages.
```
`as default for` -> `when`

kaspth commented on activesupport/lib/active_support/railtie.rb:
```diff
@@ -7,6 +7,17 @@ class Railtie < Rails::Railtie # :nodoc:
+      ActiveSupport::MessageEncryptor.use_authenticated_message_encryption = true
+    else
+      ActiveSupport::MessageEncryptor.use_authenticated_message_encryption = false
+    end
```
This all needs to be wrapped in an `initializer` (with a descriptive name) so it's run when the app boots and not when this file is parsed.

This Railtie subclass is marked as `# :nodoc` meaning it won't show up in documentation, so just skip the comments here.

I'd rewrite to this within the `initializer`:

```ruby
if config.active_support.respond_to?(:use_authenticated_message_encryption)
  ActiveSupport::MessageEncryptor.use_authenticated_message_encryption =
    config.active_support.use_authenticated_message_encryption
end
```

We have to use `respond_to?` because `config.active_support`, an instance of `ActiveSupport::OrderedOptions`, raises if the key is unassigned.

kaspth commented:
Woah, hey you got the tests to pass! 👏 — make sure you look at Code Climate too, there's one thing missing there. 😊

Now we just need to test that the default change works. Let's just go with setting `use_authenticated_message_encryption` in the Message Encryptor test and inspecting `default_cipher` for now.

Also: `default_cipher` should be marked as `def default_cipher # :nodoc:`, don't think that should be public API for people.

assain commented on activesupport/lib/active_support/railtie.rb:
```diff
@@ -7,6 +7,13 @@ class Railtie < Rails::Railtie # :nodoc:
 
     config.eager_load_namespaces << ActiveSupport
 
+    initializer "active_support.set_authenticated_message_encryption_as_default" do |app|
```
@kaspth 
Whenever, I make this change, Travis CI fails, because of sessions_test.rb which outputs:
1) Failure:
ApplicationTests::MiddlewareSessionTest#test_session_upgrading_from_AES-CBC-HMAC_encryption_to_AES-GCM_encryption [test/application/middleware/session_test.rb:351]:
Expected: "1"
  Actual: ""

kaspth replied:
Yep, see the other comment about why 😊

Let's just clip "_as_default" from the initializer title while we're here.

kaspth commented on railties/test/application/middleware/session_test.rb:
```diff
@@ -339,6 +339,9 @@ def read_raw_cookie
         config.action_dispatch.use_authenticated_cookie_encryption = true
+
+        # Enable Authenticated Message Encryption.
+        config.active_support.use_authenticated_message_encryption = true
```
This will be true by default, so there's no need to set this.

We should assign the cbc `cipher:` to the legacy encryptor in cookies middleware. Then the test passes.

kaspth commented on activesupport/lib/active_support/message_encryptor.rb:
```diff
@@ -19,7 +19,17 @@ module ActiveSupport
+    class << self
+      attr_accessor :use_authenticated_message_encryption
+
+      def default_cipher
```
Needs a `# :nodoc` same for the `use_authenticated_message_encryption` accessor.

kaspth reviewed (commented):
So close now!

kaspth commented on railties/test/application/middleware/session_test.rb:
```diff
@@ -290,7 +290,7 @@ def read_raw_cookie
       assert_equal "2", last_response.body
 
-      cipher = "aes-256-gcm"
+        
```
This shouldn't be changed.

assain replied:
Oops! My bad!

kaspth commented on railties/test/application/middleware/session_test.rb:
```diff
@@ -339,7 +339,8 @@ def read_raw_cookie
         config.action_dispatch.use_authenticated_cookie_encryption = true
-      RUBY
+
+        RUBY
```
This neither.

kaspth commented on actionpack/lib/action_dispatch/middleware/cookies.rb:
```diff
@@ -630,7 +630,7 @@ def initialize(parent_jar)
         sign_secret = key_generator.generate_key(request.encrypted_signed_cookie_salt || "")
 
-        @legacy_encryptor = ActiveSupport::MessageEncryptor.new(secret, sign_secret, digest: digest, serializer: ActiveSupport::MessageEncryptor::NullSerializer)
+        @legacy_encryptor = ActiveSupport::MessageEncryptor.new(secret, sign_secret, cipher: 'aes-256-cbc', digest: digest, serializer: ActiveSupport::MessageEncryptor::NullSerializer)
```
Use double quoted strings.
